src/Confluent.Kafka.Avro/GenericDeserializerImpl.cs

```diff
@@ -0,0 +1,94 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using Avro.IO;
+using Avro.Generic;
+using Confluent.SchemaRegistry;
+
+
+namespace Confluent.Kafka.Serialization
+{
+    internal class GenericDeserializerImpl : IAvroDeserializerImpl<GenericRecord>
+    {
+        /// <remarks>
+        ///     A datum reader cache (one corresponding to each write schema that's been seen)
+        ///     is maintained so that they only need to be constructed once.
+        /// </remarks>
+        private readonly Dictionary<int, DatumReader<GenericRecord>> datumReaderBySchemaId 
+            = new Dictionary<int, DatumReader<GenericRecord>>();
+
+        private object datumReaderLock = new object();
+
+        private ISchemaRegistryClient schemaRegistryClient;
+
+        public GenericDeserializerImpl(ISchemaRegistryClient schemaRegistryClient)
+        {
+            this.schemaRegistryClient = schemaRegistryClient;
+        }
+
+        public GenericRecord Deserialize(string topic, byte[] array)
+        {
+            // Note: topic is not necessary for deserialization (or knowing if it's a key 
+            // or value) only the schema id is needed.
+
+            using (var stream = new MemoryStream(array))
+            using (var reader = new BinaryReader(stream))
+            {
+                var magicByte = reader.ReadByte();
+                if (magicByte != Constants.MagicByte)
+                {
+                    // may change in the future.
+                    throw new InvalidDataException($"magic byte should be 0, not {magicByte}");
+                }
+                var writerIdBigEndian = reader.ReadInt32();
+                var writerId = IPAddress.NetworkToHostOrder(writerIdBigEndian);
+
+                DatumReader<GenericRecord> datumReader = null;
+
+                lock (datumReaderLock)
+                {
+                    datumReaderBySchemaId.TryGetValue(writerId, out datumReader);
+                    if (datumReader == null)
+                    {
+                        // TODO: If any of this cache fills up, this is probably an
+                        // indication of misuse of the deserializer. Ideally we would do 
+                        // something more sophisticated than the below + not allow 
+                        // the misuse to keep happening without warning.
+                        if (datumReaderBySchemaId.Count > schemaRegistryClient.MaxCachedSchemas)
+                        {
+                            datumReaderBySchemaId.Clear();
+                        }
+
+                        // we retain the lock over this blocking call since there may be concurrent
```

Comment:
I'm not sure how common multi-threaded consumption is in .NET clients, but an alternative would be to keep a separate lock on each cached schema and maintain a state (resolving, resolved) that other threads could wait on for a specific schema rather than being blocked by the entire schemaRegistryClient.

Reply:
mow you mention it, I guess this would be v. uncommon and my comment about retaining the lock over the blocking call is unimportant. The change @DavidBetteridge suggests in #449 is appropriate I think (simply use thread safe collections). I'll make that change.

Reply:
From what I understand that proposed solution will not work because of race conditions.

Reply:
I think it's ok (see note below). the cache is just trying to avoid constructing `datumReader` instances unnecessarily, it doesn't actually matter if this happens.

Reply:
there's no tech debt issue with the threading. I think not warning the user if the cache fills up is an issue (I think fail fast is better + force the user to manually flag they're doing something that is probably wrong to continue, but everyone else disagreed with me...), but I also don't think we need to track it. if people hit it, we'll find out about it and be reminded fix it.

Reply:
After giving this some more thought...

There are two scenarios:
 1. consumers are used from a single thread only - this function will only be called by one thread at any given time. No blocking.
 2. consumers are possibly used from multiple threads - this function will be called simultaneously from multiple threads, possibly deserializing messages with the same schema-id, which will result in multiple parallell requests to the schema-registry. Multiply this by the number of consumer instances, apps, etc, and there will be an unnecessary spike of HTTP requests to the schema-registry.

Now, with the two different solutions:
 a. only lock datumReaderBySchemaId (race condition)
 b. lock entire Lookup/QuerySR/Insert cycle (no race condition)

For scenario 1 there is no difference between **a** and **b**.
For scenario 2:
 * with solution **a** there will be multiple outstanding schema-registry requests which we all can agree is a bad thing in large deployments. With SR serving each request in order it means there will be no speed-up, rather the opposite, for the awaiting consumer code wanting its messages deserialized.
 * with solution **b** there will be just one schema-registry request and the other consumer threads will block waiting for the schema to be queried and inserted. The wait time of the blocked threads will be no longer than the query+insert time of the first thread.

With this in mind I strongly vote for going back to the initial implementation with no race conditions (solution b).